### ISSUE TITLE
NVDA doesn't read consent banner on page load

### ISSUE THREAD

**Issue by paruchurisilpa** [bbc/simorgh#8855]
## Reproduction Steps
1.Go to http://192.168.1.77:7080/sport/football/55699252
2.Click on down arrow to navigate to the consent banner
3.It doesn't appear to read the consent banner

Windows 10 NVDA version 2020.3 Browser ff

## Context
When accessing our pages in Firefox, using the screenreader NVDA, the consent banner loads in after NVDA begins to interact with the accessibility tree. This means that the banner is not announced as part of the page contents, and the NVDA cursor starts at the brand logo in the navigation (missing the consent banner out). 

## A11y Acceptance criteria

* If privacy banner shows, h2 should capture focus and be announced upon the banner loading in client-side
* Clicking OK button should close banner and show cookie banner
* When cookie shows, h2 should capture focus and be announced upon the banner loading in client-side
* Upon selecting Accept/Reject, the user's focus should move to the brand link at the top of the page
* There should be no focus outline on the h2 when focused.
* Users should be able to navigate out of the consent banner once focus is captured.

_This issue requires a dev a11y swarm only around this specific bug, this should include testing with keyboard only navigating via the tab key, and should also be tested with all our supported screen readers_

**Comment by andrewscfc** [comment]
Looking into this issue I took a snapshot of the contents of the `header` section of a story page:
<details>
<summary>header element html from Sport page</summary>

```html
<header role="banner">
  <div class="css-850ovu-ConsentBannerWrapper e1nbsi9l0">
    <div dir="ltr" class="css-y66d5n-Wrapper e1o5up6o0">
      <div dir="ltr" class="css-1wxd04j-CenterWrapper e1o5up6o1">
        <h2 dir="ltr" class="css-1p9wy7w-Title e1o5up6o2">
          We've updated our Privacy and Cookies Policy
        </h2>
        <p dir="ltr" class="css-l8k6jk-ConsentBannerText e1o5up6o4">
          We've made some important changes to our Privacy and Cookies Policy
          and we want you to know what this means for you and your data.
        </p>
        <ul dir="ltr" class="css-d2u9ht-Options e1o5up6o3">
          <li dir="ltr" class="css-181k8nc-ListItem e1o5up6o5">
            <button type="button">OK</button>
          </li>
          <li dir="ltr" class="css-181k8nc-ListItem e1o5up6o5">
            <a href="https://www.bbc.co.uk/usingthebbc/your-data-matters"
              >Find out what's changed</a
            >
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div class="css-n283ug-Banner e1wgjeuw1">
    <div class="css-1y4zfvh-SvgWrapper e1wgjeuw0">
      <a href="/sport" class="css-1qsa44m-StyledLink e1wgjeuw2"
        ><svg
          height="24"
          viewBox="0 0 190 24"
          xmlns="http://www.w3.org/2000/svg"
          focusable="false"
          aria-hidden="true"
          class="css-mjqao9-BrandSvg e1wgjeuw3"
        >
          <g
            fill-rule="evenodd"
            stroke="#fff"
            stroke-width=".335"
            style="stroke: #000"
          >
            <path
              d="M59.234 24h25.547V0H59.234zm6.707-18.32C67.7 4.078 70.254 3.12 73.125 3.12c1.836 0 3.754.399 6.227 1.36l.16.078v3.203l-.399-.242c-2.156-1.2-3.992-1.68-5.988-1.68-3.992 0-6.863 2.64-6.863 6.16 0 3.602 2.793 6.078 6.945 6.078 1.914 0 4.07-.637 6.066-1.84l.399-.238v3.281l-.16.078c-1.754.88-3.75 1.52-6.547 1.52-3.43 0-5.906-1.2-7.504-2.957-1.594-1.684-2.395-3.844-2.395-5.684 0-2.636 1.04-4.879 2.875-6.558M29.777 24h25.547V0H29.777zm6.707-20.64h5.426c1.996 0 3.434.562 4.39 1.44.962.802 1.438 2 1.438 3.122 0 .879-.238 1.68-.718 2.316-.399.48-.875.883-1.516 1.203 2.316.637 3.754 2.32 3.754 4.48 0 1.118-.48 2.317-1.438 3.2-1.039.879-2.636 1.52-4.87 1.52h-6.388V3.359zm9.578 12.242c0-.723-.238-1.204-.878-1.68-.559-.402-1.516-.64-2.793-.64h-2.793V18h2.953c1.437 0 2.316-.32 2.875-.8.476-.481.636-1.04.636-1.598m-2.394-5.68c.559-.402.797-.961.797-1.762 0-.558-.156-1.039-.559-1.441-.398-.399-1.117-.64-2.234-.64h-2.074v4.48h1.437c1.195 0 2.153-.239 2.633-.637M.398 24h25.547V0H.398zM7.105 3.36h5.43c1.996 0 3.43.562 4.39 1.44.958.88 1.356 2 1.356 3.122 0 .879-.238 1.68-.718 2.316-.399.48-.88.883-1.516 1.203 2.316.637 3.75 2.32 3.75 4.48 0 1.118-.477 2.317-1.434 3.2-1.039.879-2.636 1.52-4.87 1.52H7.104zm7.184 6.562c.559-.402.8-.961.8-1.762 0-.558-.16-1.039-.558-1.441-.402-.399-1.12-.64-2.238-.64h-2.074v4.48h1.437c1.196 0 2.074-.239 2.633-.637m2.395 5.68c0-.723-.239-1.204-.875-1.68-.563-.402-1.52-.64-2.797-.64h-2.793V18h2.953c1.437 0 2.316-.32 2.875-.8.476-.481.637-1.04.637-1.598m85.261 7.039c2.633 0 4.711-.563 6.149-1.68 1.437-1.121 2.152-2.723 2.152-4.64 0-2.083-.637-3.52-1.992-4.321-1.36-.8-2.957-1.441-4.871-1.922-1.117-.316-2.074-.558-2.715-.957-.637-.32-.957-.883-.957-1.68 0-1.52 1.195-2.32 3.59-2.32 2.238 0 4.234.559 5.988 1.598v-3.84c-1.754-.879-3.75-1.36-5.988-1.36-2.473 0-4.39.56-5.824 1.68-1.438 1.121-2.157 2.64-2.157 4.563 0 1.84.559 3.117 1.754 4 1.2.879 2.875 1.52 5.11 2.078 1.039.238 1.836.558 2.554.879.719.32 1.04.96 1.04 1.84 0 1.601-1.278 2.48-3.91 2.48-2.24 0-4.391-.64-6.47-1.84v3.922c2.079.957 4.153 1.52 6.547 1.52m15.965-.48v-6.88h2.957c2.555 0 4.547-.562 5.985-1.761 1.437-1.122 2.156-2.801 2.156-4.961s-.719-3.758-2.074-4.957c-1.438-1.122-3.434-1.762-5.989-1.762h-7.265v20.238h4.23zm0-16.802h2.715c2.637 0 3.914 1.121 3.914 3.282 0 2.16-1.277 3.199-3.914 3.199h-2.715zm16.207 14.403c1.836 1.918 4.23 2.879 7.344 2.879 3.113 0 5.508-.961 7.348-2.88 1.836-1.921 2.71-4.48 2.71-7.683 0-3.2-.874-5.758-2.71-7.68-1.84-1.918-4.235-2.878-7.348-2.878-3.113 0-5.508.96-7.344 2.878-1.836 1.922-2.715 4.48-2.715 7.68-.078 3.203.88 5.762 2.715 7.684m3.196-12.64c.957-1.2 2.312-1.763 4.07-1.763 1.754 0 3.113.563 4.07 1.762.957 1.2 1.438 2.8 1.438 4.957 0 2.16-.48 3.762-1.438 4.961-.957 1.2-2.316 1.762-4.07 1.762-1.758 0-3.113-.563-4.07-1.762-.961-1.2-1.438-2.879-1.438-4.96 0-2.157.477-3.759 1.438-4.958m24.984 7.758h.402a237.053 237.053 0 014.23 7.36h5.028c-.797-1.438-1.676-2.798-2.473-4.16a69.614 69.614 0 00-2.554-4c2.394-1.04 3.672-2.88 3.672-5.599 0-2.082-.72-3.68-2.157-4.8-1.437-1.121-3.433-1.68-6.148-1.68h-7.422v20.238h4.309v-7.36zm-3.113-9.52h2.875c2.714 0 4.07 1.04 4.07 3.04s-1.356 2.96-4.07 2.96h-2.793v-6zm19.879.403V22.16h4.39V5.762h6.387V2h-17.082v3.762zm0 0"
            ></path>
          </g></svg
        ><span class="css-ow3j0s-VisuallyHiddenText e1yt7oin0"
          >BBC Sport</span
        ></a
      ><a dir="ltr" href="#content" class="css-13fvji4-SkipLink e1kgr8rg0"
        >Skip to content</a
      >
    </div>
  </div>
  <nav role="navigation" dir="ltr" class="css-f6h3mv-StyledNav e1lim4kn5">
    <div class="css-h6pr6t-NavWrapper e1lim4kn0">
      <div class="css-ddfiod-ScrollableWrapper eg26bel0">
        <button
          type="button"
          aria-expanded="false"
          dir="ltr"
          class="css-13nhaas-MenuButton ex2l5ol6"
        >
          <svg
            focusable="false"
            width="2.75rem"
            height="2.75rem"
            viewBox="0 0 44 44"
            aria-hidden="true"
            class="css-15xd63c-MenuIcon eqe69jx0"
          >
            <path
              d="M12 29h21v-2.333H12V29zm0-5.833h21v-2.334H12v2.334zM12 15v2.333h21V15H12z"
            ></path></svg
          ><span class="css-ow3j0s-VisuallyHiddenText e1yt7oin0">Sections</span>
        </button>
        <div dir="ltr" class="css-138btz5-StyledScrollableNav e1fc18nj0">
          <ul role="list" class="css-1p24brr-StyledUnorderedList e1lim4kn1">
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a href="/sport" class="css-1dvwnp0-StyledLink e1lim4kn2">Home</a>
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a href="/sport/football" class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Football</a
              >
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a href="/sport/formula1" class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Formula 1</a
              >
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a href="/sport/cricket" class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Cricket</a
              >
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a
                href="/sport/rugby-union"
                class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Rugby U</a
              >
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a
                href="/sport/rugby-league"
                class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Rugby L</a
              >
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a href="/sport/tennis" class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Tennis</a
              >
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a href="/sport/golf" class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Golf</a
              >
            </li>
            <li
              dir="ltr"
              role="listitem"
              class="css-1hl65t0-StyledListItem e1lim4kn3"
            >
              <a
                href="/sport/athletics"
                class="css-1dvwnp0-StyledLink e1lim4kn2"
                >Athletics</a
              >
            </li>
          </ul>
        </div>
      </div>
      <div height="0" class="css-lmnc98-StyledDropdown ex2l5ol0">
        <ul role="list" class="css-1594wcs-DropdownUl ex2l5ol2">
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a href="/sport" class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Home</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/football"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Football</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/formula1"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Formula 1</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/cricket"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Cricket</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/rugby-union"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Rugby U</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/rugby-league"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Rugby L</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/tennis"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Tennis</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/golf"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Golf</a
            >
          </li>
          <li role="listitem" class="css-10ufcb9-StyledDropdownLi ex2l5ol3">
            <a
              href="/sport/athletics"
              class="css-1yifa9o-StyledDropdownLink ex2l5ol4"
              >Athletics</a
            >
          </li>
        </ul>
      </div>
    </div>
  </nav>
</header>
```

</details>

The consent banner currently sits within a `header` element labelled with the [`banner` role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Banner_role) mdn defining the banner role as: 
> A banner role represents general and informative content frequently placed at the beginning of the page. This usually includes a logo, company name, search icon, photo related to the page, or slogan.

I'm not sure a cookie banner fits under that description particularly as the banner is no longer visibly at the top of the page and is pinned to the bottom.

Doing a little reading I'd suggest the consent banner should be have the [`dialog` role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/dialog_role) mdn describing it as:
> The dialog role is used to mark up an HTML based application dialog or window that separates content or UI from the rest of the web application or page. Dialogs are generally placed on top of the rest of the page content using an overlay. Dialogs can be either non-modal (it's still possible to interact with content outside of the dialog) or modal (only the content in the dialog can be interacted with).

The banner is overlayed over the content and pinned to the bottom of the page so seems to fit this definition pretty neatly. I am hoping that screen readers such as NVDA may be more likely to read the content in markd up in this way? I'd be interested to hear your thoughts @greenc05 ?

**Comment by greenc05** [comment]
Is this a side note to the issue? Not sure how it would fix it?

Aside from that, a modal is something we have discussed at lengh and it's not technically possible at this time (because we are using the same code for when this is non modal) 

In my opinion this does fit the descrition (this is general and informative content that is frequendlty placed at the beginning of the page) so should be in the header landmark.

**Comment by andrewscfc** [comment]
> Is this a side note to the issue? Not sure how it would fix it?

I'm suggesting that if it was marked up more accurately (debateable I appreciate), the screen reader may be more likely to read it out. Aside from that I have no specific ideas how to go about fixing this other than marking it up in a different way?

**Comment by andrewscfc** [comment]
> Aside from that, a modal is something we have discussed at lengh and it's not technically possible at this time (because we are using the same code for when this is non modal)

The dialog role suggests it is applicable to both modal and non-modal dialogs, our banner itself being non-modal.

**Comment by greenc05** [comment]
As we've already established that this works fine in Storybook. Next I would recommend that you copy and paste the raw HTLM from the page in question that has the bug and test this with NVDA. Does the bug exist?

**Comment by greenc05** [comment]
@andrewscfc You may wish to read the GEL guide on Action dialogs (modals) https://bbc.github.io/gel/components/action-dialogs/

**Comment by greenc05** [comment]
You could also use the 'Accessiblity' tab in the inspector to try and debug on the page in question

**Comment by andrewscfc** [comment]
> As we've already established that this works fine in Storybook. Next I would recommend that you copy and paste the raw HTLM from the page in question that has the bug and test this with NVDA. Does the bug exist?

The html included in my first comment reads out in nvda fine, I just opened this file on its own in firefox
[story-banner.html.zip](https://github.com/bbc/simorgh/files/5959106/story-banner.html.zip)

I would imagine though the bug relates to the context it is used in, i.e. the full page.

**Comment by greenc05** [comment]
Could you try that with the full page HTML?

**Comment by andrewscfc** [comment]
> You could also use the 'Accessiblity' tab in the inspector to try and debug on the page in question

hmm this looks all right to my untrained eye? Can't see why nvda wouldn't read it first
<img width="961" alt="image" src="https://user-images.githubusercontent.com/9645462/107531947-453fe500-6bb5-11eb-853b-60221db55692.png">

**Comment by andrewscfc** [comment]
> Could you try that with the full page HTML?

Great suggestion, I did this by viewing the page source and the consent banner is absent when the page first loads, clientside js deciding to show it based on a cookies; I wonder if NVDA builds the structure of the document before the clientside js has chance to run?

With this new insight, I think this adds weight to marking this up as a [dialog](https://bbc.github.io/gel/components/action-dialogs/) as it is dynamic content shown based on the user's cookies. In the case of the privacy banner the information is informative and can be optionally dismissed, in the case of the cookie banner, the banner must be acted upon or left to display on each visit. In both cases though, the banner is non-modal, I don't know how that fits into the gel guidelines?

**Comment by greenc05** [comment]
A note on the way NVDA works:
“On the modern, highly dynamic web, there’s not really a reliable way to determine when a page has finished loading. We used to wait until all resources finished loading before enabling browse mode, but on really large or slow loading pages or pages with a lot of images, stylesheets, JavaScript, etc., this can mean the user has to wait an unreasonable amount of time before they can interact with the document. In some cases, it means they can never interact at all; some pages in modern web apps always report as “busy”. In contrast, a sighted user sees content as it arrives. Thus, we made the decision several years ago to just enable browse mode and say all as soon as the document got focus (which is usually when the browser thinks it has enough content to do something useful).”

Seems like NVDA just hooks in really early in loading and there’s not much we can do about it. We can use a react hook to focus the user into the banner once it loads, but on poor connections or devices with poor performance running JS that might mean that they have focus stolen once they’ve already started browsing (for all users).
